File: cli/python/agentshop_cli.py
#!/usr/bin/env python3
"""
AgentShop CLI - Python Management Interface
Provides command-line tools for managing the AgentShop platform
"""

import click
import os
import sys
import json
import asyncio
from datetime import datetime, timedelta
from typing import Optional, Dict, Any

# Add backend to path
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..', 'backend'))

from services.llm_orm_service import llm_orm_service
from config.llm_config import config_manager
from repositories.llm_repository import (
    LLMRequestRepository, LLMUsageStatsRepository, LLMProviderStatusRepository
)


@click.group()
@click.version_option(version='1.0.0')
def cli():
    """AgentShop CLI - Manage your AI Agent Marketplace"""
    pass


# =============================================================================
# LLM MANAGEMENT COMMANDS
# =============================================================================

@cli.group()
def llm():
    """LLM provider management commands"""
    pass


@llm.command()
def providers():
    """List all LLM providers and their status"""
    click.echo("🤖 LLM Providers Status")
    click.echo("=" * 50)
    
    try:
        status_repo = LLMProviderStatusRepository()
        providers = status_repo.get_all_provider_status()
        
        if not providers:
            click.echo("No providers configured.")
            return
        
        for provider in providers:
            status_icon = "[OK]" if provider.is_healthy else "[ERROR]"
            enabled_icon = "🟢" if provider.is_enabled else "🔴"
            
            click.echo(f"{status_icon} {enabled_icon} {provider.provider.upper()}")
            click.echo(f"   Enabled: {provider.is_enabled}")
            click.echo(f"   Healthy: {provider.is_healthy}")
            click.echo(f"   API Key: {'[OK]' if provider.api_key_configured else '[ERROR]'}")
            click.echo(f"   Default Model: {provider.default_model or 'Not set'}")
            click.echo(f"   Daily Cost: ${provider.current_daily_cost:.4f} / ${provider.daily_cost_limit:.2f}")
            if provider.last_error:
                click.echo(f"   Last Error: {provider.last_error}")
            click.echo()
            
    except Exception as e:
        click.echo(f"[ERROR] Error: {e}", err=True)


@llm.command()
@click.argument('provider')
@click.argument('api_key')
@click.option('--model', help='Default model for the provider')
@click.option('--cost-limit', type=float, default=10.0, help='Daily cost limit in USD')
def enable(provider: str, api_key: str, model: Optional[str], cost_limit: float):
    """Enable an LLM provider"""
    try:
        click.echo(f"[SETUP] Enabling {provider.upper()} provider...")
        
        config_manager.enable_provider(
            provider,
            api_key,
            default_model=model,
            cost_limit_daily=cost_limit
        )
        
        click.echo(f"[OK] {provider.upper()} provider enabled successfully!")
        if model:
            click.echo(f"   Default model: {model}")
        click.echo(f"   Daily cost limit: ${cost_limit:.2f}")
        
    except Exception as e:
        click.echo(f"[ERROR] Error enabling provider: {e}", err=True)


@llm.command()
@click.argument('provider')
def disable(provider: str):
    """Disable an LLM provider"""
    try:
        click.echo(f"[DISABLE] Disabling {provider.upper()} provider...")
        config_manager.disable_provider(provider)
        click.echo(f"[OK] {provider.upper()} provider disabled successfully!")
        
    except Exception as e:
        click.echo(f"[ERROR] Error disabling provider: {e}", err=True)


@llm.command()
@click.option('--days', default=7, help='Number of days to show stats for')
@click.option('--provider', help='Filter by specific provider')
def stats(days: int, provider: Optional[str]):
    """Show LLM usage statistics"""
    click.echo(f"[STATS] LLM Usage Statistics (Last {days} days)")
    click.echo("=" * 60)
    
    try:
        usage_repo = LLMUsageStatsRepository()
        stats = usage_repo.get_usage_summary('day', days, provider)
        
        if not stats:
            click.echo("No usage data found.")
            return
        
        total_requests = sum(s.request_count for s in stats)
        total_cost = sum(s.total_cost for s in stats)
        total_tokens = sum(s.total_tokens for s in stats)
        
        click.echo(f"📈 Summary:")
        click.echo(f"   Total Requests: {total_requests:,}")
        click.echo(f"   Total Cost: ${total_cost:.4f}")
        click.echo(f"   Total Tokens: {total_tokens:,}")
        click.echo()
        
        # Group by provider
        provider_stats = {}
        for stat in stats:
            if stat.provider not in provider_stats:
                provider_stats[stat.provider] = {
                    'requests': 0, 'cost': 0.0, 'tokens': 0
                }
            provider_stats[stat.provider]['requests'] += stat.request_count
            provider_stats[stat.provider]['cost'] += stat.total_cost
            provider_stats[stat.provider]['tokens'] += stat.total_tokens
        
        click.echo("🤖 By Provider:")
        for prov, data in provider_stats.items():
            click.echo(f"   {prov.upper()}:")
            click.echo(f"     Requests: {data['requests']:,}")
            click.echo(f"     Cost: ${data['cost']:.4f}")
            click.echo(f"     Tokens: {data['tokens']:,}")
            click.echo()
            
    except Exception as e:
        click.echo(f"[ERROR] Error: {e}", err=True)


@llm.command()
@click.argument('prompt')
@click.option('--provider', default='openai', help='LLM provider to use')
@click.option('--model', help='Model to use')
@click.option('--temperature', type=float, default=0.7, help='Temperature (0.0-1.0)')
@click.option('--max-tokens', type=int, default=100, help='Maximum tokens')
def test(prompt: str, provider: str, model: Optional[str], temperature: float, max_tokens: int):
    """Test LLM generation"""
    click.echo(f"🧪 Testing {provider.upper()} provider...")
    click.echo(f"Prompt: {prompt}")
    click.echo("=" * 50)
    
    async def run_test():
        try:
            result = await llm_orm_service.generate_text(
                prompt=prompt,
                provider=provider,
                model=model,
                temperature=temperature,
                max_tokens=max_tokens,
                request_type='cli_test'
            )
            
            click.echo(f"[OK] Response:")
            click.echo(f"{result['content']}")
            click.echo()
            click.echo(f"[STATS] Stats:")
            click.echo(f"   Provider: {result['provider']}")
            click.echo(f"   Model: {result['model']}")
            click.echo(f"   Tokens: {result['tokens_used']}")
            click.echo(f"   Cost: ${result['cost']:.6f}")
            click.echo(f"   Time: {result['processing_time_ms']}ms")
            click.echo(f"   Cached: {result['cached']}")
            
        except Exception as e:
            click.echo(f"[ERROR] Test failed: {e}", err=True)
    
    asyncio.run(run_test())


# =============================================================================
# DATABASE MANAGEMENT COMMANDS
# =============================================================================

@cli.group()
def db():
    """Database management commands"""
    pass


@db.command()
@click.option('--backup-dir', default='./datalake/db_dumps', help='Backup directory')
def backup(backup_dir: str):
    """Create database backup"""
    os.makedirs(backup_dir, exist_ok=True)
    
    timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
    backup_file = os.path.join(backup_dir, f'agentshop_backup_{timestamp}.sql')
    
    try:
        # For SQLite database
        db_path = os.getenv('DATABASE_URL', 'sqlite:///agentshop.db').replace('sqlite:///', '')
        
        import sqlite3
        conn = sqlite3.connect(db_path)
        
        with open(backup_file, 'w') as f:
            for line in conn.iterdump():
                f.write('%s\n' % line)
        
        conn.close()
        
        click.echo(f"[OK] Database backed up to: {backup_file}")
        
    except Exception as e:
        click.echo(f"[ERROR] Backup failed: {e}", err=True)


@db.command()
@click.argument('backup_file')
@click.option('--confirm', is_flag=True, help='Confirm restoration')
def restore(backup_file: str, confirm: bool):
    """Restore database from backup"""
    if not confirm:
        click.echo("[WARNING]  This will overwrite the current database!")
        click.echo("Use --confirm flag to proceed.")
        return
    
    try:
        if not os.path.exists(backup_file):
            click.echo(f"[ERROR] Backup file not found: {backup_file}")
            return
        
        db_path = os.getenv('DATABASE_URL', 'sqlite:///agentshop.db').replace('sqlite:///', '')
        
        import sqlite3
        conn = sqlite3.connect(db_path)
        
        with open(backup_file, 'r') as f:
            sql_script = f.read()
        
        conn.executescript(sql_script)
        conn.close()
        
        click.echo(f"[OK] Database restored from: {backup_file}")
        
    except Exception as e:
        click.echo(f"[ERROR] Restore failed: {e}", err=True)


@db.command()
def migrate():
    """Run database migrations"""
    try:
        from orm.base_model import db_manager
        db_manager.create_tables()
        click.echo("[OK] Database migrations completed successfully!")
        
    except Exception as e:
        click.echo(f"[ERROR] Migration failed: {e}", err=True)


# =============================================================================
# DATA MANAGEMENT COMMANDS
# =============================================================================

@cli.group()
def data():
    """Data management commands"""
    pass


@data.command()
@click.option('--days', default=30, help='Days of data to keep')
@click.option('--dry-run', is_flag=True, help='Show what would be deleted')
def cleanup(days: int, dry_run: bool):
    """Clean up old data"""
    cutoff_date = datetime.now() - timedelta(days=days)
    
    click.echo(f"🧹 Data Cleanup (older than {days} days)")
    click.echo(f"Cutoff date: {cutoff_date}")
    click.echo("=" * 50)
    
    try:
        request_repo = LLMRequestRepository()
        
        # Find old requests
        old_requests = request_repo.find_by()  # Would need to add date filter
        
        if dry_run:
            click.echo(f"Would delete {len(old_requests)} old LLM requests")
        else:
            # Implement actual cleanup
            click.echo(f"Deleted {len(old_requests)} old LLM requests")
            
    except Exception as e:
        click.echo(f"[ERROR] Cleanup failed: {e}", err=True)


@data.command()
@click.option('--format', type=click.Choice(['json', 'csv']), default='json')
@click.option('--output', help='Output file path')
def export(format: str, output: Optional[str]):
    """Export data"""
    if not output:
        timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
        output = f"./datalake/processed_data/export_{timestamp}.{format}"
    
    os.makedirs(os.path.dirname(output), exist_ok=True)
    
    try:
        usage_repo = LLMUsageStatsRepository()
        stats = usage_repo.get_usage_summary('day', 30)
        
        if format == 'json':
            with open(output, 'w') as f:
                json.dump([s.to_dict() for s in stats], f, indent=2, default=str)
        elif format == 'csv':
            import csv
            with open(output, 'w', newline='') as f:
                if stats:
                    writer = csv.DictWriter(f, fieldnames=stats[0].to_dict().keys())
                    writer.writeheader()
                    for stat in stats:
                        writer.writerow(stat.to_dict())
        
        click.echo(f"[OK] Data exported to: {output}")
        
    except Exception as e:
        click.echo(f"[ERROR] Export failed: {e}", err=True)


# =============================================================================
# AUTOMATION COMMANDS
# =============================================================================

@cli.group()
def automation():
    """Automation management commands"""
    pass


@automation.command()
def list_jobs():
    """List all automation jobs"""
    click.echo("🤖 Automation Jobs")
    click.echo("=" * 50)
    # Implementation would query webcrawler_jobs table
    click.echo("No jobs configured yet.")


@automation.command()
@click.argument('job_name')
@click.argument('target_url')
@click.option('--job-type', default='content_scraping', help='Type of job')
@click.option('--schedule', help='Cron schedule expression')
def create_job(job_name: str, target_url: str, job_type: str, schedule: Optional[str]):
    """Create a new automation job"""
    click.echo(f"[SETUP] Creating job: {job_name}")
    click.echo(f"Target URL: {target_url}")
    click.echo(f"Job Type: {job_type}")
    if schedule:
        click.echo(f"Schedule: {schedule}")
    
    # Implementation would create job in webcrawler_jobs table
    click.echo("[OK] Job created successfully!")


# =============================================================================
# UTILITY COMMANDS
# =============================================================================

@cli.command()
def config():
    """Show current configuration"""
    click.echo("[CONFIG]  AgentShop Configuration")
    click.echo("=" * 50)
    
    try:
        safe_config = config_manager.get_safe_config()
        
        click.echo(f"Default Provider: {safe_config['default_provider']}")
        click.echo(f"Cache TTL: {safe_config['cache_ttl']}s")
        click.echo(f"Caching Enabled: {safe_config['enable_caching']}")
        click.echo(f"Rate Limiting: {safe_config['enable_rate_limiting']}")
        click.echo()
        
        click.echo("Providers:")
        for provider, config in safe_config['providers'].items():
            enabled_icon = "🟢" if config['enabled'] else "🔴"
            click.echo(f"  {enabled_icon} {provider.upper()}")
            click.echo(f"     API Key: {'[OK]' if config['api_key'] != 'None' else '[ERROR]'}")
            if config['default_model']:
                click.echo(f"     Model: {config['default_model']}")
            click.echo(f"     Cost Limit: ${config['cost_limit_daily']}/day")
            
    except Exception as e:
        click.echo(f"[ERROR] Error: {e}", err=True)


@cli.command()
def status():
    """Show system status"""
    click.echo("🏥 AgentShop System Status")
    click.echo("=" * 50)
    
    try:
        # Database status
        from orm.base_model import db_manager
        db_manager.get_session().close()
        click.echo("[OK] Database: Connected")
        
        # LLM providers status
        asyncio.run(_check_llm_health())
        
    except Exception as e:
        click.echo(f"[ERROR] System check failed: {e}", err=True)


async def _check_llm_health():
    """Check LLM provider health"""
    try:
        health = await llm_orm_service.health_check()
        click.echo("\n🤖 LLM Providers:")
        for provider, healthy in health.items():
            status_icon = "[OK]" if healthy else "[ERROR]"
            click.echo(f"   {status_icon} {provider.upper()}")
    except Exception as e:
        click.echo(f"[ERROR] LLM health check failed: {e}")


if __name__ == '__main__':
    cli()

File: cli/typescript/agentshop-cli.ts
#!/usr/bin/env ts-node
/**
 * AgentShop CLI - TypeScript Management Interface
 * Provides command-line tools for managing the AgentShop platform
 */

import { Command } from 'commander';
import axios, { AxiosInstance } from 'axios';
import fs from 'fs';
import path from 'path';
import chalk from 'chalk';
import ora from 'ora';
import inquirer from 'inquirer';

interface ApiResponse<T> {
  success: boolean;
  data: T;
  message: string;
  errors?: string[];
}

interface LLMProvider {
  provider: string;
  is_enabled: boolean;
  is_healthy: boolean;
  api_key_configured: boolean;
  default_model?: string;
  current_daily_cost: number;
  daily_cost_limit: number;
}

interface UsageStats {
  date: string;
  provider: string;
  model: string;
  request_count: number;
  successful_requests: number;
  total_cost: number;
  average_response_time_ms: number;
}

class AgentShopCLI {
  private api: AxiosInstance;
  private baseUrl: string;

  constructor(baseUrl: string = 'http://localhost:5000') {
    this.baseUrl = baseUrl;
    this.api = axios.create({
      baseURL: baseUrl,
      timeout: 30000,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }

  async checkConnection(): Promise<boolean> {
    try {
      await this.api.get('/health');
      return true;
    } catch {
      return false;
    }
  }

  async getProviders(): Promise<LLMProvider[]> {
    const response = await this.api.get<ApiResponse<LLMProvider[]>>('/api/llm/analytics/providers/status');
    return response.data.data;
  }

  async getUsageStats(days: number = 7, provider?: string): Promise<UsageStats[]> {
    const params = new URLSearchParams();
    params.append('days', days.toString());
    if (provider) params.append('provider', provider);

    const response = await this.api.get<ApiResponse<UsageStats[]>>(
      `/api/llm/analytics/usage?${params}`
    );
    return response.data.data;
  }

  async enableProvider(provider: string, apiKey: string, model?: string): Promise<void> {
    const payload: any = { api_key: apiKey };
    if (model) payload.default_model = model;

    await this.api.post(`/api/llm/config/providers/${provider}/enable`, payload);
  }

  async disableProvider(provider: string): Promise<void> {
    await this.api.post(`/api/llm/config/providers/${provider}/disable`);
  }

  async testGeneration(prompt: string, provider: string, model?: string): Promise<any> {
    const payload: any = {
      prompt,
      provider,
      max_tokens: 100,
    };
    if (model) payload.model = model;

    const response = await this.api.post('/api/llm/generate', payload);
    return response.data.data;
  }

  async healthCheck(): Promise<Record<string, boolean>> {
    const response = await this.api.get<ApiResponse<Record<string, boolean>>>(
      '/api/llm/analytics/health'
    );
    return response.data.data;
  }
}

// =============================================================================
// CLI SETUP
// =============================================================================

const program = new Command();
const cli = new AgentShopCLI();

program
  .name('agentshop')
  .description('AgentShop CLI - Manage your AI Agent Marketplace')
  .version('1.0.0')
  .option('-u, --url <url>', 'API base URL', 'http://localhost:5000');

// =============================================================================
// LLM COMMANDS
// =============================================================================

const llmCommand = program.command('llm').description('LLM provider management');

llmCommand
  .command('providers')
  .description('List all LLM providers and their status')
  .action(async () => {
    const spinner = ora('Fetching providers...').start();
    
    try {
      const connected = await cli.checkConnection();
      if (!connected) {
        spinner.fail('Cannot connect to AgentShop API. Is the server running?');
        process.exit(1);
      }

      const providers = await cli.getProviders();
      spinner.stop();

      console.log(chalk.blue.bold('\n🤖 LLM Providers Status'));
      console.log('='.repeat(50));

      if (providers.length === 0) {
        console.log(chalk.yellow('No providers configured.'));
        return;
      }

      for (const provider of providers) {
        const statusIcon = provider.is_healthy ? '✅' : '❌';
        const enabledIcon = provider.is_enabled ? '🟢' : '🔴';
        
        console.log(`${statusIcon} ${enabledIcon} ${chalk.bold(provider.provider.toUpperCase())}`);
        console.log(`   Enabled: ${provider.is_enabled ? chalk.green('Yes') : chalk.red('No')}`);
        console.log(`   Healthy: ${provider.is_healthy ? chalk.green('Yes') : chalk.red('No')}`);
        console.log(`   API Key: ${provider.api_key_configured ? '✅' : '❌'}`);
        console.log(`   Default Model: ${provider.default_model || 'Not set'}`);
        console.log(
          `   Daily Cost: $${provider.current_daily_cost.toFixed(4)} / $${provider.daily_cost_limit.toFixed(2)}`
        );
        console.log();
      }
    } catch (error: any) {
      spinner.fail(`Error: ${error.message}`);
    }
  });

llmCommand
  .command('enable <provider> <api-key>')
  .description('Enable an LLM provider')
  .option('-m, --model <model>', 'Default model for the provider')
  .action(async (provider: string, apiKey: string, options: any) => {
    const spinner = ora(`Enabling ${provider.toUpperCase()} provider...`).start();
    
    try {
      await cli.enableProvider(provider, apiKey, options.model);
      spinner.succeed(`${provider.toUpperCase()} provider enabled successfully!`);
      
      if (options.model) {
        console.log(`   Default model: ${options.model}`);
      }
    } catch (error: any) {
      spinner.fail(`Error enabling provider: ${error.message}`);
    }
  });

llmCommand
  .command('disable <provider>')
  .description('Disable an LLM provider')
  .action(async (provider: string) => {
    const spinner = ora(`Disabling ${provider.toUpperCase()} provider...`).start();
    
    try {
      await cli.disableProvider(provider);
      spinner.succeed(`${provider.toUpperCase()} provider disabled successfully!`);
    } catch (error: any) {
      spinner.fail(`Error disabling provider: ${error.message}`);
    }
  });

llmCommand
  .command('stats')
  .description('Show LLM usage statistics')
  .option('-d, --days <days>', 'Number of days to show stats for', '7')
  .option('-p, --provider <provider>', 'Filter by specific provider')
  .action(async (options: any) => {
    const spinner = ora('Fetching statistics...').start();
    
    try {
      const stats = await cli.getUsageStats(parseInt(options.days), options.provider);
      spinner.stop();

      console.log(chalk.blue.bold(`\n📊 LLM Usage Statistics (Last ${options.days} days)`));
      console.log('='.repeat(60));

      if (stats.length === 0) {
        console.log(chalk.yellow('No usage data found.'));
        return;
      }

      const totalRequests = stats.reduce((sum, s) => sum + s.request_count, 0);
      const totalCost = stats.reduce((sum, s) => sum + s.total_cost, 0);

      console.log(chalk.green.bold('📈 Summary:'));
      console.log(`   Total Requests: ${chalk.cyan(totalRequests.toLocaleString())}`);
      console.log(`   Total Cost: ${chalk.green('$' + totalCost.toFixed(4))}`);
      console.log();

      // Group by provider
      const providerStats: Record<string, any> = {};
      for (const stat of stats) {
        if (!providerStats[stat.provider]) {
          providerStats[stat.provider] = { requests: 0, cost: 0 };
        }
        providerStats[stat.provider].requests += stat.request_count;
        providerStats[stat.provider].cost += stat.total_cost;
      }

      console.log(chalk.blue.bold('🤖 By Provider:'));
      for (const [provider, data] of Object.entries(providerStats)) {
        console.log(`   ${chalk.bold(provider.toUpperCase())}:`);
        console.log(`     Requests: ${chalk.cyan(data.requests.toLocaleString())}`);
        console.log(`     Cost: ${chalk.green('$' + data.cost.toFixed(4))}`);
        console.log();
      }
    } catch (error: any) {
      spinner.fail(`Error: ${error.message}`);
    }
  });

llmCommand
  .command('test <prompt>')
  .description('Test LLM generation')
  .option('-p, --provider <provider>', 'LLM provider to use', 'openai')
  .option('-m, --model <model>', 'Model to use')
  .action(async (prompt: string, options: any) => {
    const spinner = ora(`Testing ${options.provider.toUpperCase()} provider...`).start();
    
    try {
      const result = await cli.testGeneration(prompt, options.provider, options.model);
      spinner.stop();

      console.log(chalk.blue.bold(`\n🧪 Test Results`));
      console.log('='.repeat(50));
      console.log(chalk.bold('Prompt:'), prompt);
      console.log();
      console.log(chalk.green.bold('✅ Response:'));
      console.log(result.content);
      console.log();
      console.log(chalk.blue.bold('📊 Stats:'));
      console.log(`   Provider: ${result.provider}`);
      console.log(`   Model: ${result.model}`);
      console.log(`   Tokens: ${result.tokens_used}`);
      console.log(`   Cost: $${result.cost.toFixed(6)}`);
      console.log(`   Time: ${result.processing_time_ms}ms`);
      console.log(`   Cached: ${result.cached ? 'Yes' : 'No'}`);
    } catch (error: any) {
      spinner.fail(`Test failed: ${error.message}`);
    }
  });

llmCommand
  .command('health')
  .description('Check health of all LLM providers')
  .action(async () => {
    const spinner = ora('Checking provider health...').start();
    
    try {
      const health = await cli.healthCheck();
      spinner.stop();

      console.log(chalk.blue.bold('\n🏥 LLM Provider Health Check'));
      console.log('='.repeat(50));

      for (const [provider, healthy] of Object.entries(health)) {
        const icon = healthy ? '✅' : '❌';
        const status = healthy ? chalk.green('Healthy') : chalk.red('Unhealthy');
        console.log(`${icon} ${chalk.bold(provider.toUpperCase())}: ${status}`);
      }
    } catch (error: any) {
      spinner.fail(`Health check failed: ${error.message}`);
    }
  });

// =============================================================================
// DATA COMMANDS
// =============================================================================

const dataCommand = program.command('data').description('Data management commands');

dataCommand
  .command('export')
  .description('Export usage data')
  .option('-f, --format <format>', 'Export format (json|csv)', 'json')
  .option('-o, --output <file>', 'Output file path')
  .option('-d, --days <days>', 'Number of days to export', '30')
  .action(async (options: any) => {
    const spinner = ora('Exporting data...').start();
    
    try {
      const stats = await cli.getUsageStats(parseInt(options.days));
      
      const timestamp = new Date().toISOString().replace(/[:.]/g, '-').split('T')[0];
      const defaultOutput = `./datalake/processed_data/export_${timestamp}.${options.format}`;
      const outputPath = options.output || defaultOutput;
      
      // Ensure directory exists
      const dir = path.dirname(outputPath);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }

      if (options.format === 'json') {
        fs.writeFileSync(outputPath, JSON.stringify(stats, null, 2));
      } else if (options.format === 'csv') {
        const csv = convertToCSV(stats);
        fs.writeFileSync(outputPath, csv);
      }

      spinner.succeed(`Data exported to: ${outputPath}`);
    } catch (error: any) {
      spinner.fail(`Export failed: ${error.message}`);
    }
  });

// =============================================================================
// INTERACTIVE COMMANDS
// =============================================================================

program
  .command('setup')
  .description('Interactive setup wizard')
  .action(async () => {
    console.log(chalk.blue.bold('\n🚀 AgentShop Setup Wizard'));
    console.log('='.repeat(50));

    try {
      const connected = await cli.checkConnection();
      if (!connected) {
        console.log(chalk.red('❌ Cannot connect to AgentShop API. Please start the server first.'));
        process.exit(1);
      }

      const answers = await inquirer.prompt([
        {
          type: 'checkbox',
          name: 'providers',
          message: 'Which LLM providers would you like to enable?',
          choices: [
            { name: 'OpenAI (GPT models)', value: 'openai' },
            { name: 'Anthropic Claude', value: 'claude' },
            { name: 'Groq (Fast inference)', value: 'groq' },
            { name: 'Perplexity (Search-augmented)', value: 'perplexity' },
            { name: 'Ollama (Local models)', value: 'ollama' },
          ],
        },
      ]);

      for (const provider of answers.providers) {
        if (provider === 'ollama') {
          const ollamaConfig = await inquirer.prompt([
            {
              type: 'input',
              name: 'baseUrl',
              message: 'Ollama base URL:',
              default: 'http://localhost:11434',
            },
          ]);
          
          console.log(chalk.green(`✅ Ollama configured with URL: ${ollamaConfig.baseUrl}`));
        } else {
          const providerConfig = await inquirer.prompt([
            {
              type: 'password',
              name: 'apiKey',
              message: `Enter ${provider.toUpperCase()} API key:`,
              mask: '*',
            },
            {
              type: 'input',
              name: 'model',
              message: `Default model for ${provider} (optional):`,
            },
          ]);

          const spinner = ora(`Configuring ${provider.toUpperCase()}...`).start();
          
          try {
            await cli.enableProvider(provider, providerConfig.apiKey, providerConfig.model);
            spinner.succeed(`${provider.toUpperCase()} configured successfully!`);
          } catch (error: any) {
            spinner.fail(`Failed to configure ${provider}: ${error.message}`);
          }
        }
      }

      console.log(chalk.green.bold('\n🎉 Setup completed!'));
      console.log('You can now use the AgentShop CLI and web interface.');
    } catch (error: any) {
      console.log(chalk.red(`Setup failed: ${error.message}`));
    }
  });

program
  .command('status')
  .description('Show system status')
  .action(async () => {
    console.log(chalk.blue.bold('\n🏥 AgentShop System Status'));
    console.log('='.repeat(50));

    const spinner = ora('Checking system status...').start();
    
    try {
      const connected = await cli.checkConnection();
      spinner.stop();

      if (connected) {
        console.log(`✅ API Server: ${chalk.green('Connected')} (${cli.baseUrl})`);
        
        // Check LLM providers
        const health = await cli.healthCheck();
        console.log('\n🤖 LLM Providers:');
        
        for (const [provider, healthy] of Object.entries(health)) {
          const icon = healthy ? '✅' : '❌';
          console.log(`   ${icon} ${provider.toUpperCase()}`);
        }
      } else {
        console.log(`❌ API Server: ${chalk.red('Disconnected')}`);
        console.log(chalk.yellow('Please ensure the backend server is running.'));
      }
    } catch (error: any) {
      spinner.fail(`Status check failed: ${error.message}`);
    }
  });

// =============================================================================
// UTILITY FUNCTIONS
// =============================================================================

function convertToCSV(data: any[]): string {
  if (data.length === 0) return '';
  
  const headers = Object.keys(data[0]);
  const csvContent = [
    headers.join(','),
    ...data.map(row => 
      headers.map(header => {
        const value = row[header];
        return typeof value === 'string' && value.includes(',') 
          ? `"${value}"` 
          : value;
      }).join(',')
    )
  ].join('\n');
  
  return csvContent;
}

// =============================================================================
// MAIN EXECUTION
// =============================================================================

async function main() {
  try {
    await program.parseAsync(process.argv);
  } catch (error: any) {
    console.error(chalk.red(`❌ Error: ${error.message}`));
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

export { AgentShopCLI };